tutorials/haskell/mustache-templates/tutorial.md

```diff
@@ -0,0 +1,404 @@
+
+Let's see how to compile templates because that's where you usually start.
+But first, let's take a look at types that will show up here and there in
+the rest of the tutorial:
+
+* `PName` stands for “partial's name”. “Partials” are other templates that
+  are inserted into actual template you're rendering. `PName` is a wrapper
+  around `Text` and defined to make it harder to mix it up with other
+  textual values. If you enable the `OverloadedStrings` extension, you can
+  write `PName`s just as normal `String`s.
+
+* `Node` is a piece of template. The whole template body is represented as
+  `[Node]` and that's all you need to know unless you plan to manipulate
+  parsed representation of template, which is also easy, but not in the
+  scope of this tutorial.
+
+### Three ways to compile a template
+
+A `Template` is actually a collection of templates (partials) with one of
+them selected:
+
+```haskell
+-- | Mustache template as name of “top-level” template and a collection of
+-- all available templates (partials).
+--
+-- 'Template' is a 'Semigroup'. This means that you can combine 'Template's
+-- (and their caches) using the ('<>') operator, the resulting 'Template'
+-- will have the same currently selected template as the left one. Union of
+-- caches is also left-biased.
+
+data Template = Template
+  { templateActual :: PName
+    -- ^ Name of currently “selected” template (top-level one).
+  , templateCache  :: Map PName [Node]
+    -- ^ Collection of all templates that are available for interpolation
+    -- (as partials). The top-level one is also contained here and the
+    -- “focus” can be switched easily by modifying 'templateActual'.
+  } deriving (Eq, Ord, Show, Data, Typeable, Generic)
+```
+
+When working with a `Template`, it's only possible to use partials  that are
+in `templateCache`, thus the main difference between compiling functions is
+where you get your template(s) and what you will have in `templateCache`.
+Thus there are three different ways to get a `Template`:
+
+1. From lazy `Text` with `compileMustacheText`. The function takes a `PName`
+   and actual template source. The function returns `Either` parse error or
+   `Template` with cache consisting of only one `Template`.
+
+2. From a single `File` with `compileMustacheFile`. This one only takes path
+   to Mustache file the resulting template won't be able to use partials
+   unless you combine several templates into one (see below).
+
+3. From a directory with `compileMustacheDir`. This function reads all
+   templates (files ending with `.mustache` extension) and puts them into
+   the cache, selecting one of them as main (you specify which). The
+   resulting `Template` can use partials that were present in that
+   directory.
+
+So to use partials, we need to use `compileMustacheDir` *or* combine several
+templates into one using `(<>)` method of `Semigroup` type class. Being an
+instance of the `Semigroup` type class means that you can always to two
```

Comment:
@mrkkrp "combine two" or "append two" (or something other than "to"). Above ("*or*"), remember that the current font we have doesn't have an italic version.